Added "Maximum holding time" setting to services and analyses

xispa opened the pull request:
## Description of the issue/feature this PR addresses

This Pull Request allows the user to set the maximum holding time before it does no longer make sense to perform a given test. In the context of sample analysis, maximum holding time refers to the maximum duration a sample can be stored after collection before it must be analyzed to ensure reliable and accurate results. This period is established to prevent degradation, contamination, or chemical changes that could alter the sample's composition

For instance, testing for arboviral illnesses like Dengue, Zika, and Chikungunya might not make sense 5 days after the collection date. Likewise, identification of pathogenic microorganisms in a food sample won't make sense e.g. 2d after collection.

A new field "Maximum holding time" has been added to services and analyses:

![Captura de 2024-10-31 17-01-56](https://github.com/user-attachments/assets/c6034a38-93af-4f65-a250-f812af8c6183)

On sample creation form, when the user selects a collection date that makes not possible to conduct a given analysis within the holding time, the service becomes not available for selection and an icon is displayed instead, along with an informative text:

![Captura de 2024-10-08 14-12-17](https://github.com/user-attachments/assets/d8cb249b-0036-474b-946d-2043190b6ecd)

Service is still available for selection from "Manage analyses" view though.

An alert icon is also displayed in analyses listing either when the result was captured past the holding time limit or when the holding time is about to expire:

![Captura de 2024-10-10 10-58-46](https://github.com/user-attachments/assets/c4530260-de73-486e-902d-b0cc8dc057b0)


--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

xispa committed:
Added analytical holding time functionality

xispa committed:
Typo

xispa committed:
Redux

xispa committed:
Trigger recalculate_records on sample date change

xispa committed:
do not display the lock button if beyond holding time

xispa committed:
Remove unnecessary bits

Surprisingly, this snippet is not necessary because when a sample
is created via add form, the uids of services (those assigned to
profiles included) are passed on submit

xispa committed:
Add warning icon when analysis is beyond holding time

xispa committed:
Remove unnecessary bits

xispa committed:
Clean import

xispa committed:
Changelog

xispa committed:
Merge branch '2.x' of github.com:senaite/senaite.core into hold-time

xispa committed:
Merge branch '2.x' into hold-time

xispa committed:
Improve texts

xispa committed:
Better wording

xispa committed:
CHangelog

xispa committed:
better spelling

ramonski committed:
Merge branch '2.x' into hold-time

xispa committed:
Merge branch '2.x' into hold-time

xispa committed:
Merge branch '2.x' into hold-time

ramonski committed:
Merge branch '2.x' into hold-time

ramonski reviewed (commented):
Cool stuff!

I played around with the Date Sampled field after the sample was created with an Analysis contained that had a max holding time of 1 day set. After I changed to the date of today, I got the following traceback:

```
Traceback (innermost last):
  Module ZPublisher.WSGIPublisher, line 176, in transaction_pubevents
  Module ZPublisher.WSGIPublisher, line 385, in publish_module
  Module ZPublisher.WSGIPublisher, line 288, in publish
  Module ZPublisher.mapply, line 85, in mapply
  Module ZPublisher.WSGIPublisher, line 63, in call_object
  Module senaite.app.listing.view, line 246, in __call__
  Module senaite.app.listing.ajax, line 113, in handle_subpath
  Module senaite.core.decorators, line 40, in decorator
  Module senaite.app.listing.decorators, line 63, in wrapper
  Module senaite.app.listing.decorators, line 50, in wrapper
  Module senaite.app.listing.decorators, line 100, in wrapper
  Module senaite.app.listing.ajax, line 383, in ajax_folderitems
  Module senaite.app.listing.decorators, line 88, in wrapper
  Module senaite.app.listing.ajax, line 259, in get_folderitems
  Module bika.lims.browser.analyses.view, line 808, in folderitems
  Module senaite.app.listing.view, line 982, in folderitems
  Module bika.lims.browser.analyses.view, line 797, in folderitem
  Module bika.lims.browser.analyses.view, line 1758, in _folder_item_holding_time
TypeError: can't compare datetime.datetime to str
```

<img width="1533" alt="" src="https://github.com/user-attachments/assets/2c137ca4-3e22-4ce4-9cc4-807649f16383">

ramonski commented on src/bika/lims/utils/analysisrequest.py:
```diff
@@ -267,14 +266,6 @@ def to_list(value):
-        for brain in uid_catalog(UID=profiles):
-            profile = api.get_object(brain)
-            uids.extend(profile.getServiceUIDs() or [])
-
```
Great to have this removed. I see that this is already handled by the `setProfiles` method in the sample content.

ramonski reviewed (commented):
I tested it with profiles and templates and it worked perfectly.
The only issue seem to be when changing afterwards the Date sampled to today in the sample that has such an Analysis Service added

xispa committed:
Fix TypeError: can't compare datetime.datetime to str

xispa commented:
> Cool stuff!
> 
> I played around with the Date Sampled field after the sample was created with an Analysis contained that had a max holding time of 1 day set. After I changed to the date of today, I got the following traceback:
> 
> ```
> Traceback (innermost last):
>   Module ZPublisher.WSGIPublisher, line 176, in transaction_pubevents
>   Module ZPublisher.WSGIPublisher, line 385, in publish_module
>   Module ZPublisher.WSGIPublisher, line 288, in publish
>   Module ZPublisher.mapply, line 85, in mapply
>   Module ZPublisher.WSGIPublisher, line 63, in call_object
>   Module senaite.app.listing.view, line 246, in __call__
>   Module senaite.app.listing.ajax, line 113, in handle_subpath
>   Module senaite.core.decorators, line 40, in decorator
>   Module senaite.app.listing.decorators, line 63, in wrapper
>   Module senaite.app.listing.decorators, line 50, in wrapper
>   Module senaite.app.listing.decorators, line 100, in wrapper
>   Module senaite.app.listing.ajax, line 383, in ajax_folderitems
>   Module senaite.app.listing.decorators, line 88, in wrapper
>   Module senaite.app.listing.ajax, line 259, in get_folderitems
>   Module bika.lims.browser.analyses.view, line 808, in folderitems
>   Module senaite.app.listing.view, line 982, in folderitems
>   Module bika.lims.browser.analyses.view, line 797, in folderitem
>   Module bika.lims.browser.analyses.view, line 1758, in _folder_item_holding_time
> TypeError: can't compare datetime.datetime to str
> ```
> 
> <img alt="" width="1533" src="https://private-user-images.githubusercontent.com/713193/382914186-2c137ca4-3e22-4ce4-9cc4-807649f16383.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA4MDM2NzMsIm5iZiI6MTczMDgwMzM3MywicGF0aCI6Ii83MTMxOTMvMzgyOTE0MTg2LTJjMTM3Y2E0LTNlMjItNGNlNC05Y2M0LTgwNzY0OWYxNjM4My5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTA1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEwNVQxMDQyNTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jODcwNjE0ZGZhZDQwY2UwMmY1YjE2ZWI5MzFmMDdlM2JkZDZjZGU3Y2RjMGY2MTgzN2M0OGM4YjIxOGFkYTVkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Ce2t88WFTx1UQEU_FWc57tG_R9nBu8kBFQPC3_P0Yqg">

Fixed with https://github.com/senaite/senaite.core/pull/2624/commits/48e763101145fd199c8767a18d2be432db904efd

Note I use ANSIs to compare dates. Is useful for sorting and for preventing the infamous `TypeError: can't compare offset-naive and offset-aware datetimes` when comparing a TZ-naive datetime with a non-naive.

However, I now realized that this might lead to inconsistencies when the TZ of the collection date (DateSampled) is different from system's TZ.

To overccome this problem I am considering to add the param `to_uc=False` in `to_ansi` function, so it always tries to convert to Universal Time before formatting. This saves the developer from all dance re TZ conversion and naive cases before `to_ansi` is called. Or even better, I can use the parameter `tz=None` instead, so if `tz` is set, the system will try to convert to that TZ. Umm let me try this.

xispa committed:
Added timezone parameter to `to_ansi` dtime function

xispa committed:
Add now() function in api.dtime

xispa committed:
Merge branch '2.x' of github.com:senaite/senaite.core into hold-time

xispa committed:
Fix doctest

xispa committed:
Fix inconsistency with Etc/GMT timezone when datetime is used

xispa committed:
Rely on to_ansi with timezone for comparisons

xispa committed:
Fix UnknownTimeZoneError when extracting the timezone from datetime

```
Traceback (innermost last):
  Module ZPublisher.WSGIPublisher, line 176, in transaction_pubevents
  Module ZPublisher.WSGIPublisher, line 385, in publish_module
  Module ZPublisher.WSGIPublisher, line 288, in publish
  Module ZPublisher.mapply, line 85, in mapply
  Module ZPublisher.WSGIPublisher, line 63, in call_object
  Module senaite.app.listing.view, line 246, in __call__
  Module senaite.app.listing.ajax, line 113, in handle_subpath
  Module senaite.core.decorators, line 40, in decorator
  Module senaite.app.listing.decorators, line 63, in wrapper
  Module senaite.app.listing.decorators, line 50, in wrapper
  Module senaite.app.listing.decorators, line 100, in wrapper
  Module senaite.app.listing.ajax, line 383, in ajax_folderitems
  Module senaite.app.listing.decorators, line 88, in wrapper
  Module senaite.app.listing.ajax, line 259, in get_folderitems
  Module bika.lims.browser.analyses.view, line 808, in folderitems
  Module senaite.app.listing.view, line 982, in folderitems
  Module bika.lims.browser.analyses.view, line 797, in folderitem
  Module bika.lims.browser.analyses.view, line 1737, in _folder_item_holding_time
  Module senaite.core.api.dtime, line 262, in to_ansi
  Module senaite.core.api.dtime, line 382, in to_zone
  Module pytz, line 188, in timezone
UnknownTimeZoneError: '+02'
```

xispa committed:
Do not raise error if timezone is not valid in to_ansi

xispa committed:
Better wording

xispa committed:
Proper comparison of dates in sample add form

xispa commented:
> Note I use ANSIs to compare dates. Is useful for sorting and for preventing the infamous `TypeError: can't compare offset-naive and offset-aware datetimes` when comparing a TZ-naive datetime with a non-naive.
> 
> However, I now realized that this might lead to inconsistencies when the TZ of the collection date (DateSampled) is different from system's TZ.
> 
> To overccome this problem I am considering to add the param `to_uc=False` in `to_ansi` function, so it always tries to convert to Universal Time before formatting. This saves the developer from all dance re TZ conversion and naive cases before `to_ansi` is called. Or even better, I can use the parameter `tz=None` instead, so if `tz` is set, the system will try to convert to that TZ. Umm let me try this.

Pheew... had to go through `api.dtime` and resolve some inconsistencies with TZ conversions. Hope it becomes cleaner now.

xispa committed:
Merge branch '2.x' into hold-time

ramonski committed:
Merge branch '2.x' into hold-time

ramonski reviewed (approved):
Works like a charm, many thanks for this!
